tex/polyglossia.sty

```diff
@@ -1262,34 +1252,53 @@
 }
 
 \newcommand\xpg@ifdefined[3]{%
-    \ifluatex
-        \xpg@set@language@luatex@iv{#1}%
-    \fi
+    \xpg@set@hyphenation@patterns{#1}%
     \polyglossia@check@ifdefined:NTF{#1}{#2}{#3}
 }%
 
-\newcommand\xpg@set@language@luatex@iv[1]{%
-    \ifluatex
-        \polyglossia@check@ifdefined:NF{#1}{
+% Set \bbl@hyphendata@\the\language, which is (lua)babel's
+% hyphenation pattern hook
+% FIXME Clarifiy why/when this is needed.
+\newcommand*\xpg@set@bbl@hyphendata[1]{%
+    \ifluatex%
+        \ifcsdef{bbl@hyphendata@#1}{}{%
+            \global\@namedef{bbl@hyphendata@\the\language}{}%
+        }%
+    \fi% 
+}
+
+% Set hyphenation patterns for a given language. This does the right
+% thing both for XeTeX and LuaTeX
+\newcommand\xpg@set@hyphenation@patterns[1]{%
+    \str_case_e:nnF{\c_sys_engine_str}{%
+      {luatex}{%
+        \polyglossia@check@ifdefined:NF{#1}{%
             \expandafter\chardef\csname l@#1\endcsname=\directlua{tex.sprint(polyglossia.newloader('#1'))}%
-        }
-        \language=\csname l@#1\endcsname
-    \fi
+            \language=\csname l@#1\endcsname%
```

Comment:
Originally this `\language=...` line was outside of `\polyglossia@check@ifdefined:NF` conditional block, but now it is inside. This seems to be wrong, becase language will not be declared properly except the first call of `\xpg@set@hyphenation@patterns` macro. 
Again, the same line in the xetex part below should be outside conditional block.

Reply:
Could you post a MWE where this fails?

Reply:
```latex
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage{french}
\begin{document}
\showthe\language
la française.
\loggingoutput
\end{document}
```
Upon lualatex, originally the number shown by the line 5 was `2`, and we could see a discretionary node in the log file:
```
....\TU/lmr/m/n/10 f
....\TU/lmr/m/n/10 r
....\TU/lmr/m/n/10 a
....\TU/lmr/m/n/10 n
....\discretionary (penalty 50)
.....< \TU/lmr/m/n/10 -
....\TU/lmr/m/n/10 ç
....\TU/lmr/m/n/10 a
....\TU/lmr/m/n/10 i
....\TU/lmr/m/n/10 s
....\TU/lmr/m/n/10 e
```
After applying this PR, however, the number is now `0` (USenglish) and so we see no discretionary in the log file.

Reply:
OK, thanks. Should be cured.